internal/actions/update.go

```diff
@@ -51,128 +67,122 @@ func Update(client container.Client, params types.UpdateParams) (types.Report, e
 	} else {
-		failedStop, stoppedImages := stopContainersInReversedOrder(containersToUpdate, client, params)
-		progress.UpdateFailed(failedStop)
-		failedStart := restartContainersInSortedOrder(containersToUpdate, client, params, stoppedImages)
-		progress.UpdateFailed(failedStart)
+		stoppedImages := us.stopContainersInReversedOrder(containersToUpdate)
+		us.restartContainersInSortedOrder(containersToUpdate, stoppedImages)
 	}
 
-	if params.LifecycleHooks {
-		lifecycle.ExecutePostChecks(client, params)
-	}
-	return progress.Report(), nil
+	return nil
 }
 
-func performRollingRestart(containers []types.Container, client container.Client, params types.UpdateParams) map[types.ContainerID]error {
+func (us *updateSession) performRollingRestart(containers []types.Container) {
 	cleanupImageIDs := make(map[types.ImageID]bool, len(containers))
 	failed := make(map[types.ContainerID]error, len(containers))
 
 	for i := len(containers) - 1; i >= 0; i-- {
 		if containers[i].ToRestart() {
-			err := stopStaleContainer(containers[i], client, params)
+			err := us.stopContainer(containers[i])
 			if err != nil {
 				failed[containers[i].ID()] = err
 			} else {
-				if err := restartStaleContainer(containers[i], client, params); err != nil {
+				if err := us.restartContainer(containers[i]); err != nil {
 					failed[containers[i].ID()] = err
-				} else if containers[i].IsStale() {
+				} else if containers[i].IsMarkedForUpdate() {
 					// Only add (previously) stale containers' images to cleanup
 					cleanupImageIDs[containers[i].ImageID()] = true
 				}
 			}
 		}
 	}
 
-	if params.Cleanup {
-		cleanupImages(client, cleanupImageIDs)
+	if us.params.Cleanup {
+		us.cleanupImages(cleanupImageIDs)
 	}
-	return failed
+	us.progress.UpdateFailed(failed)
 }
 
-func stopContainersInReversedOrder(containers []types.Container, client container.Client, params types.UpdateParams) (failed map[types.ContainerID]error, stopped map[types.ImageID]bool) {
-	failed = make(map[types.ContainerID]error, len(containers))
+func (us *updateSession) stopContainersInReversedOrder(containers []types.Container) (stopped map[types.ImageID]bool) {
+	failed := make(map[types.ContainerID]error, len(containers))
 	stopped = make(map[types.ImageID]bool, len(containers))
 	for i := len(containers) - 1; i >= 0; i-- {
-		if err := stopStaleContainer(containers[i], client, params); err != nil {
+		if err := us.stopContainer(containers[i]); err != nil {
 			failed[containers[i].ID()] = err
 		} else {
 			// NOTE: If a container is restarted due to a dependency this might be empty
 			stopped[containers[i].SafeImageID()] = true
 		}
 
 	}
-	return
+	us.progress.UpdateFailed(failed)
+
+	return stopped
 }
 
-func stopStaleContainer(container types.Container, client container.Client, params types.UpdateParams) error {
-	if container.IsWatchtower() {
-		log.Debugf("This is the watchtower container %s", container.Name())
+func (us *updateSession) stopContainer(c types.Container) error {
```

Comment:
Consider keeping this variable named "container" rather than "c" to help in readability and to match other functions like restartContainer

Reply:
Can't, since that is the name of an imported package. This is the common go "solution". :/